### DISCUSSION TITLE
Problems with pydantic mapping and sqlalchemy model based on query

### DISCUSSION THREAD

**Discussion by rv2931** [fastapi/fastapi#12409] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import APIRouter
from pydantic import BaseModel
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    Double,
    ForeignKey,
    Integer,
    Interval,
    String,
    PrimaryKeyConstraint
)
from sqlalchemy import select, func
from sqlalchemy.orm import mapped_column, Mapped, relationship
from sqlalchemy import create_engine, exc, orm
from sqlalchemy.orm import Session

engine = create_engine("sqlite:///mapping.db", echo=True)

#==============================================================================
# SQLALCHEMY MODELS
#==============================================================================
BaseEntity = declarative_base()

class PersonEntity(BaseEntity):
    __tablename__ = "person"
    id = Column("id", Integer, primary_key=True)
    name = Column("name", String)
class RideEntity(BaseEntity):
    __tablename__ = "ride"
    id = Column("id", Integer, primary_key=True)
    distance = Column("distance_km", Double)
    person_id = Column("person_id", Integer, ForeignKey("person.id"), nullable=False)


total_ride_per_person_request=(
                       select(
                           PersonEntity.id,
                           RideEntity.person_id,
                           func.sum(RideEntity.distance).label("total_distance")
                        )\
                .select_from(RideEntity)\
                .join(PersonEntity, PersonEntity.id == RideEntity.person_id)\
                .group_by(PersonEntity.id,RideEntity.person_id)\
                .subquery())

class TotalRidePerPersonEntity(BaseEntity):
    __table__ = total_ride_per_person_request
    person : Mapped[PersonEntity] = relationship()
    total_distance: Mapped[Double]
#==============================================================================
# PYDANTIC SCHEMAS
#==============================================================================

class PersonSchema(BaseModel):
    id: int
    name: str

class RideSchema(BaseModel):
    id: int
    distance: float
    person_id: int

class TotalRidePerPersonSchema(BaseModel):
    class Config:
        from_attributes: bool = True
    person: PersonSchema
    total_distance: float


#==============================================================================
# FASTAPI ROUTER
#==============================================================================
router = APIRouter()
@router.get("/mapping",
            response_model=list[TotalRidePerPersonSchema],
            tags=['Test'])
async def test_mapping():
    with Session(engine) as session:
        person = PersonEntity(id = 0, name="John")
        ride1 = RideEntity(id = 0, distance=30.0, person_id=0)
        ride2 = RideEntity(id = 1, distance=20.0, person_id=0)

        session.add_all([person,ride1, ride2])
        session.commit()
        result = session.execute(select(TotalRidePerPersonEntity)).all()
        return result
BaseEntity.metadata.create_all(engine)
```

### Description

Hello
I want to deal with SQLAlchemy model and Pydantic Schema with a SQLAlchemy model not based on a table but on query as described in the example code

My problème is with this code I get a ResponseValidationError  from pydantic
```
[...]
fastapi.exceptions.ResponseValidationError: 2 validation errors:
  {'type': 'missing', 'loc': ('response', 0, 'person'), 'msg': 'Field required', 'input': (<bloom.routers.mapping.TotalRidePerPersonEntity object at 0x7fee6a5200d0>,)}
  {'type': 'missing', 'loc': ('response', 0, 'total_distance'), 'msg': 'Field required', 'input': (<bloom.routers.mapping.TotalRidePerPersonEntity object at 0x7fee6a5200d0>,)}
```
I just access to endpoint http://localhost:8000/mapping to call endpoint code and get the error

It seems that the mapping between the list of list[TotalRidePerPersonEntity] as result is not correctly done with the response_model=list[TotalRidePerPersonSchema] from pydantic

I don't find the reason
If I display the content of session.execute(select(TotalRidePerPersonEntity)).scalars() I get TotalRidePerPersonEntity objet with correct values for each fields...
Does anyone can tell what I'm doing wrong ?
Thank you in advance

### Operating System

Linux, Windows

### Operating System Details

working with Podman on Windows in a debian container/python:3.10

### FastAPI Version

0.111.0

### Pydantic Version

2.6.4

### Python Version

3.10

**Comment by deight93** [comment, marked as answer]
Hello!!

After testing the code above, it seems there are two issues.


First,

before : 
`result = session.execute(select(TotalRidePerPersonEntity)).all()`
This code returns a list of tuples from the query result. Each tuple contains the selected entity, so the TotalRidePerPersonEntity object is inside the tuple.
`result print -> [(<main.TotalRidePerPersonEntity object at 0x1104120d0>,)]`

after : 
`result = session.execute(select(TotalRidePerPersonEntity)).scalars().all()`
This code uses ScalarResult to return a list of entity objects directly. The scalars() method extracts the first element from each tuple, making it more suitable for integration with FastAPI when you have a response_model like list[TotalRidePerPersonSchema]
`result print -> [<main.TotalRidePerPersonEntity object at 0x1104120d0>]`


Second,

When using `session.execute(select(TotalRidePerPersonEntity)).scalars().all()`, a DetachedInstanceError occurs due to the lazy loading nature of the relationship():

```
fastapi.exceptions.ResponseValidationError: 1 validation errors:
  {'type': 'get_attribute_error', 'loc': ('response', 0, 'person'), 'msg': "Error extracting attribute: DetachedInstanceError: Parent instance <TotalRidePerPersonEntity at 0x1104120d0> is not bound to a Session; lazy load operation of attribute 'person' cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3)", 'input': <main.TotalRidePerPersonEntity object at 0x1104120d0>, 'ctx': {'error': "DetachedInstanceError: Parent instance <TotalRidePerPersonEntity at 0x1104120d0> is not bound to a Session; lazy load operation of attribute 'person' cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3)"}
```

This can be resolved by changing the loading strategy to various eager loading techniques.

example :
`person : Mapped[PersonEntity] = relationship(lazy='joined')`
`result = session.execute(select(TotalRidePerPersonEntity).options(selectinload(TotalRidePerPersonEntity.person))).scalars().all()`
`result = session.execute(select(TotalRidePerPersonEntity).options(joinedload(TotalRidePerPersonEntity.person))).scalars().all()`
...


Reference:
https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#relationship-loading-techniques

**Reply by rv2931** [reply]
Okay... Not so simple
My first question is what the advantage or inconvenience of working with scalars instead of tuples ? I imagine that tuples have some reason for SQLAlchemy. I've seen that select multiple models in the select is not compliant with scalars as it extracts only the first object of tuple so not the second. To workaround this, I think (maybe I'm wrong) that it is possible to create a "metamodel" that takes the two models in one (query model ie)
But what am I losing working with scalars ?

The other way I've seen and I managed to do it with more flexibility is to create its own mapper functions to avoid "automatic mapping" of fast API based on response_model. This solutions looks a little harder but allow to do what you want in mapping, having multiple schema (according to admin permission or not for example).
So I'm wondering if I choose to make my own mappers OR trying to conform to fast API constraints using scalars with maybe loosing for some Sqlalchemy facilities and it's tuples
